src/main/webapp/authorization.js

```diff
@@ -1,56 +1,115 @@
-  const firebase = getFirebase();
-  if (firebase.apps.length === 0) {
-    firebase.initializeApp(config);
+class GithubAuthorizer {
+  constructor() { // eslint-disable-line
+    GithubAuthorizer.initializeFirebase();
+    this.token = null;
+    this.firebase = firebase; // eslint-disable-line
   }
-}
 
-/**
- * Get firebase
- * @return {Firebase}
- */
-function getFirebase() {
-  return firebase; //eslint-disable-line
-}
+  /**
+   * The most recent access token for the Github API. To refresh this, the user
+   * must be signed out and signed in again. It is important to keep in mind
+   * that this token expires and must be used swiftly after calling
+   * toggleSignIn(). If authorization is yet to be toggled, this will return
+   * null. If an error occurs during authorization, it will again be set to
+   * null. This property will be set to null on sign out as well.
+   *
+   * NOTE: In the event that an access token times out, the token will still be
+   * available here and so, a non-null token is not a good sign of a valid
+   * token.
+   *
+   * @return {string} most recent token or null
+   */
+  getToken() {
+    return this.token;
+  }
 
+  /**
+   * the current user if authorized and null if not. For all intents and
+   * purposes, this function will serve as a true/false value in an if
+   * statement.
+   *
+   * @return {Firebase.User} current user or null
+   */
+  getUser() {
+    return this.firebase.auth().currentUser;
+  }
 
-/**
- * Function called when clicking the Login/Logout button.
- */
-function toggleSignIn() {
-  const firebase = getFirebase();
-  if (!firebase.auth().currentUser) {
-    let provider = new firebase.auth.GithubAuthProvider();
-    firebase.auth().signInWithPopup(provider).then(function(result) {
-      // This gives you a GitHub Access Token.
-      // You can use it to access the GitHub API.
-      const token = result.credential.accessToken; //eslint-disable-line
-      // The signed-in user info.
-      const user = result.user; //eslint-disable-line
-    }).catch(function(error) {
-      // Handle Errors here.
-      console.error(error);
-      // Change this in future iterations
-      alert(error.message);
-    });
-  } else {
-    // if logged in, sign out
-    firebase.auth().signOut();
+  /**
+   * Get firebase instance used with this object
+   * @return {Firebase} firebase instance
+   */
+  getFirebase() {
+    return this.firebase;
+  }
+
+  /**
+   * Function called when authorizing user to verify them as a github user.
+   * This function doubles as an opt out when the current user wants to leave
+   * our site for good.
+   *
+   * Side Effects: when authorizing a user, it sets the current objects token
+   * to the returned Github API accessToken if a token is returned.
+   *
+   * NOTE: this function is asynchronous and should be used with an await
+   *
+   * @return {string} the GitHub API access token, null if an error occurs or
+   * the given user is no longer athorized.
```

Comment:
I'm wondering if it'd be better to return a boolean from this function that indicates whether or not the toggle was successful. Otherwise, it becomes a bit awkward to verify that the toggle was successful, as a null value is returned in the case of both an error and a successful signout.

Reply:
Yeah that makes a lot of sense.